parser/prism/Translator.cc

```diff
@@ -5057,7 +5057,25 @@ NodeVec Translator::translateEnsure(pm_begin_node *beginNode) {
 
             auto loc = translateLoc(start, end);
 
-            translatedEnsure = make_unique<parser::Ensure>(loc, move(translatedRescue), move(ensureBody));
+            if (!directlyDesugar || !hasExpr(translatedRescue, ensureBody)) {
+                translatedEnsure = make_unique<parser::Ensure>(loc, move(translatedRescue), move(ensureBody));
+            } else {
+                // Build ast::Rescue expression with ensure field set
+                // When we have both rescue and ensure, the translatedRescue is already an ast::Rescue,
+                // so we just need to set its ensure field
+                auto bodyExpr = translatedRescue->takeDesugaredExpr();
+                auto rescue = ast::cast_tree<ast::Rescue>(bodyExpr);
```

Comment:
```suggestion
                auto rescue = ast::NodeWithExpr::cast_tree<ast::Rescue>(bodyExpr);
```

(I don't think it makes a difference, but I'd rather use the `NodeWithExpr` variant throughout the whole file, so we don't have to mentally manage a list of acceptable exceptions vs not.)